### PR TITLE
Fix aggregation memory revoke when operator is finishing

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 1]
```diff
@@ -155,8 +155,7 @@ private boolean hasPreviousSpillCompletedSuccessfully()
     public ListenableFuture<?> startMemoryRevoke()
     {
         checkState(spillInProgress.isDone());
-        spillToDisk();
-        return spillInProgress;
+        return spillToDisk();
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: also remove `checkState(spillInProgress.isDone());` above (this should be, and it is, checked in `spillToDisk` method)

File: presto-main/src/test/java/io/prestosql/operator/TestHashAggregationOperator.java [hunk 2]
```diff
@@ -154,6 +154,8 @@ public void tearDown()
     @Test(dataProvider = "hashEnabledAndMemoryLimitForMergeValues")
     public void testHashAggregation(boolean hashEnabled, boolean spillEnabled, long memoryLimitForMerge, long memoryLimitForMergeWithMemory)
     {
+        // make operator produce multiple pages during finish phase
+        int numberOfRows = 40000;
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/operator/TestHashAggregationOperator.java]: nit: 40_000

File: presto-main/src/test/java/io/prestosql/operator/TestHashAggregationOperator.java [hunk 3]
```diff
@@ -164,9 +166,9 @@ public void testHashAggregation(boolean hashEnabled, boolean spillEnabled, long
         List<Integer> hashChannels = Ints.asList(1);
         RowPagesBuilder rowPagesBuilder = rowPagesBuilder(hashEnabled, hashChannels, VARCHAR, VARCHAR, VARCHAR, BIGINT, BOOLEAN);
         List<Page> input = rowPagesBuilder
-                .addSequencePage(10, 100, 0, 100, 0, 500)
-                .addSequencePage(10, 100, 0, 200, 0, 500)
-                .addSequencePage(10, 100, 0, 300, 0, 500)
+                .addSequencePage(numberOfRows, 100, 0, 100000, 0, 500)
+                .addSequencePage(numberOfRows, 100, 0, 200000, 0, 500)
+                .addSequencePage(numberOfRows, 100, 0, 300000, 0, 500)
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/operator/TestHashAggregationOperator.java]: nit: 300_000

File: presto-main/src/test/java/io/prestosql/operator/TestHashAggregationOperator.java [hunk 4]
```diff
@@ -196,20 +198,22 @@ public void testHashAggregation(boolean hashEnabled, boolean spillEnabled, long
...
+        }
+
+        List<Page> pages = toPages(operatorFactory, driverContext, input);
+        assertGreaterThan(pages.size(), 0, "Expected more then one output page");
+        if (hashEnabled) {
+            // Drop the hashChannel for all pages for comparison
+            pages = dropChannel(pages, hashChannels);
+        }
+
+        MaterializedResult expected = expectedBuilder.build();
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/operator/TestHashAggregationOperator.java]: move up, just after the `for` loop which populates `expectedBuilder`

File: presto-main/src/test/java/io/prestosql/operator/TestHashAggregationOperator.java [hunk 5]
```diff
@@ -196,20 +198,22 @@ public void testHashAggregation(boolean hashEnabled, boolean spillEnabled, long
...
+        MaterializedResult actual = toMaterializedResult(driverContext.getSession(), expected.getTypes(), pages);
+        assertEqualsIgnoreOrder(actual.getMaterializedRows(), expected.getMaterializedRows());
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/operator/TestHashAggregationOperator.java]: I don't get it. Previously the code was calling `assertOperatorEqualsIgnoreOrder`
Now it's more verbose, including manually calling `toPages`, dropping hash channels, converting to `MaterializedResult`.
i assume this is `assertOperatorEqualsIgnoreOrder` inlined, but what's the actual difference?

Is this really related to increasing the data set, as `Produce more than single page in testHashAggregation` commit message says?

File: presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 6]
```diff
@@ -173,6 +188,21 @@ private boolean shouldMergeWithMemory(long memorySize)
     public WorkProcessor<Page> buildResult()
     {
         checkState(hasPreviousSpillCompletedSuccessfully(), "Previous spill hasn't yet finished");
+        producingOutput = true;
+
+        // convert revocable memory to user memory as memory must not be revoked when aggregation operator is finishing
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: This isn't really explaining why. Either shorten this to
```
// Convert revocable memory to user memory
```
or add some better explanation. Perhaps (if i got this right):
```
// Convert revocable memory to user memory as returned WorkProcessor holds on to memory so we no longer can revoke.
```

File: presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 7]
```diff
@@ -173,6 +188,21 @@ private boolean shouldMergeWithMemory(long memorySize)
...
+        if (localRevocableMemoryContext.getBytes() > 0) {
+            long currentRevocableBytes = localRevocableMemoryContext.getBytes();
+            localRevocableMemoryContext.setBytes(0);
+            if (!localUserMemoryContext.trySetBytes(hashAggregationBuilder.getSizeInMemory())) {
+                // TODO: this might fail (even though we have just released memory), but we don't
+                // have a proper way to atomically convert memory reservations
+                localRevocableMemoryContext.setBytes(currentRevocableBytes);
+                // spill since revocable memory could not be converted to user memory immediately
+                getFutureValue(spillToDisk());
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: This can take time.
Can we spill asynchronously?
```
spillToDisk();
return WorkProcessor.empty();
```
(then we need to `updateMemory();` on next call..)

File: presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 8]
```diff
@@ -173,6 +188,21 @@ private boolean shouldMergeWithMemory(long memorySize)
...
+        producingOutput = true;
+
+        // convert revocable memory to user memory as memory must not be revoked when aggregation operator is finishing
+        if (localRevocableMemoryContext.getBytes() > 0) {
+            long currentRevocableBytes = localRevocableMemoryContext.getBytes();
+            localRevocableMemoryContext.setBytes(0);
+            if (!localUserMemoryContext.trySetBytes(hashAggregationBuilder.getSizeInMemory())) {
+                // TODO: this might fail (even though we have just released memory), but we don't
+                // have a proper way to atomically convert memory reservations
+                localRevocableMemoryContext.setBytes(currentRevocableBytes);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: for `trySetBytes` you were passing `hashAggregationBuilder.getSizeInMemory()`, here `currentRevocableBytes`.

i would use `currentRevocableBytes + localUserMemoryContext.getBytes()` in both places, so that the code in this method does not need to know how memory footpring is calculated. Let's leave this to `updateMemory()`.

File: presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 9]
```diff
@@ -155,6 +164,12 @@ private boolean hasPreviousSpillCompletedSuccessfully()
     public ListenableFuture<?> startMemoryRevoke()
     {
         checkState(spillInProgress.isDone());
+
+        if (producingOutput) {
+            // all revocable memory has been released in buildResult method
+            return NOT_BLOCKED;
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: `verify(localRevocableMemoryContext.getBytes() == 0)`

File: presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 10]
```diff
@@ -112,8 +114,15 @@ public SpillableHashAggregationBuilder(
     public void updateMemory()
     {
         checkState(spillInProgress.isDone());
-        localUserMemoryContext.setBytes(emptyHashAggregationBuilderSize);
-        localRevocableMemoryContext.setBytes(hashAggregationBuilder.getSizeInMemory() - emptyHashAggregationBuilderSize);
+
+        if (producingOutput) {
+            localRevocableMemoryContext.setBytes(0);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: `verify(localRevocableMemoryContext.getBytes() == 0)` instead

File: presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 11]
```diff
@@ -173,6 +188,21 @@ private boolean shouldMergeWithMemory(long memorySize)
...
         checkState(hasPreviousSpillCompletedSuccessfully(), "Previous spill hasn't yet finished");
+        producingOutput = true;
+
+        // convert revocable memory to user memory as memory must not be revoked when aggregation operator is finishing
+        if (localRevocableMemoryContext.getBytes() > 0) {
+            long currentRevocableBytes = localRevocableMemoryContext.getBytes();
+            localRevocableMemoryContext.setBytes(0);
+            if (!localUserMemoryContext.trySetBytes(hashAggregationBuilder.getSizeInMemory())) {
+                // TODO: this might fail (even though we have just released memory), but we don't
+                // have a proper way to atomically convert memory reservations
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: `localRevocableMemoryContext.setBytes(currentRevocableBytes);` might fail for other reasons as well -- revocable reservation can be above query max memory per node and we are going over the limit here. Would it be possible to spill before calling `localRevocableMemoryContext.setBytes(currentRevocableBytes);` ?

File: presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java (hunk 6 above)
Comment (sopel39) [reply presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: The latter is nicer

File: presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java (hunk 7 above)
Comment (sopel39) [reply presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: Added todo since this requires bigger operator refactor

**Comment by sopel39** [comment]
ac

File: presto-main/src/test/java/io/prestosql/operator/TestHashAggregationOperator.java [hunk 12]
```diff
@@ -196,20 +199,15 @@ public void testHashAggregation(boolean hashEnabled, boolean spillEnabled, long
...
-                .row("9", 3L, 27L, 9.0, "309", 3L, 3L)
-                .build();
+        MaterializedResult.Builder expectedBuilder = resultBuilder(driverContext.getSession(), VARCHAR, BIGINT, BIGINT, DOUBLE, VARCHAR, BIGINT, BIGINT);
+        for (int i = 0; i < numberOfRows; ++i) {
+            expectedBuilder.row(Integer.toString(i), 3L, 3L * i, (double) i, Integer.toString(300_000 + i), 3L, 3L);
+        }
+        MaterializedResult expected = expectedBuilder.build();
+
+        List<Page> pages = toPages(operatorFactory, driverContext, input);
+        assertPagesEqualIgnoreOrder(driverContext, pages, expected, hashEnabled, Optional.of(hashChannels.size()));
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/operator/TestHashAggregationOperator.java]: also add `assertThat(pages.size()).isGreaterThan(1);` or something like that
